codegen-serde/src/main/kotlin/software/amazon/smithy/rust/codegen/serde/SerializeImplGenerator.kt

```diff
@@ -0,0 +1,500 @@
+import software.amazon.smithy.rust.codegen.core.smithy.contextName
+import software.amazon.smithy.rust.codegen.core.smithy.generators.UnionGenerator
+import software.amazon.smithy.rust.codegen.core.smithy.generators.renderUnknownVariant
+import software.amazon.smithy.rust.codegen.core.smithy.isOptional
+import software.amazon.smithy.rust.codegen.core.smithy.protocols.shapeFunctionName
+import software.amazon.smithy.rust.codegen.core.smithy.protocols.shapeModuleName
+import software.amazon.smithy.rust.codegen.core.smithy.rustType
+import software.amazon.smithy.rust.codegen.core.util.PANIC
+import software.amazon.smithy.rust.codegen.core.util.dq
+import software.amazon.smithy.rust.codegen.core.util.hasTrait
+import software.amazon.smithy.rust.codegen.core.util.isTargetUnit
+import software.amazon.smithy.rust.codegen.core.util.letIf
+import software.amazon.smithy.rust.codegen.core.util.toPascalCase
+import software.amazon.smithy.rust.codegen.server.smithy.hasConstraintTrait
+
+class SerializeImplGenerator(private val codegenContext: CodegenContext) {
+    private val model = codegenContext.model
+
+    fun generateRootSerializerForShape(shape: Shape): Writable = serializerFn(shape, null)
+
+    /**
+     * Generates a serializer for a given shape. Collection serializer require using a wrapper structure. To handle this,
+     * `applyTo` allows passing in a `Writable` referring to an input. This returns a writable that wraps the input if necessary,
+     * e.g.
+     *
+     * `&self.map_shape` => `SerializeSomeSpecificMap(&self.map_shape)`
+     *
+     * Even if no wrapping is required, the writable that is returned includes the required serializer as dependencies.
+     */
+    private fun serializerFn(
+        shape: Shape,
+        applyTo: Writable?,
+    ): Writable {
+        if (shape is ServiceShape) {
+            return shape.operations.map { serializerFn(model.expectShape(it), null) }.join("\n")
+        } else if (shape is OperationShape) {
+            return writable {
+                serializerFn(model.expectShape(shape.inputShape), null)(this)
+                serializerFn(model.expectShape(shape.outputShape), null)(this)
+            }
+        }
+        val name = codegenContext.symbolProvider.shapeFunctionName(codegenContext.serviceShape, shape) + "_serde"
+        val deps =
+            when (shape) {
+                is StructureShape -> RuntimeType.forInlineFun(name, serdeSubmodule(shape), structSerdeImpl(shape))
+                is UnionShape -> RuntimeType.forInlineFun(name, serdeSubmodule(shape), serializeUnionImpl(shape))
+                is TimestampShape -> serializeDateTime(shape)
+                is BlobShape ->
+                    if (shape.hasTrait<StreamingTrait>()) {
+                        serializeByteStream(shape)
+                    } else {
+                        serializeBlob(shape)
+                    }
+
+                is StringShape, is NumberShape, is BooleanShape -> directSerde(shape)
```

Comment:
These shapes could be constrained and as such we can't just delegate to `directSerde`.

Reply:
yeah correct. I've cleaned this up.